src/Microsoft.ServiceFabric.Data.Indexing.Persistent/FilterableIndex.cs

```diff
@@ -30,41 +38,118 @@ public FilterableIndex(string name, Func<TKey, TValue, TFilter> filter)
+        {
+            string name = valuePropertyName ?? throw new ArgumentNullException(nameof(valuePropertyName));
+            if (typeof(TValue).GetProperty(valuePropertyName) == null)
+            {
+                throw new ArgumentException("Dictionary value type: " + typeof(TValue).ToString() + " does not have property: " + valuePropertyName + ", note that nestled types (e.g. Value/Name/FirstName) are not currently supported");
+            }
+
+            // Construct the lambda we expect the index we are looking for to have
+            // (k, v) => v.PropertyName
+            ParameterExpression keyParameterExpression = Expression.Parameter(typeof(TKey), "k");
+            ParameterExpression valueParameterExpression = Expression.Parameter(typeof(TValue), "v");
+            MemberExpression propertyMemberExpression = Expression.Property(valueParameterExpression, typeof(TValue).GetProperty(valuePropertyName));
+            LambdaExpression filterLambda = Expression.Lambda(propertyMemberExpression, new ParameterExpression[] { keyParameterExpression, valueParameterExpression });
+
+            return new FilterableIndex<TKey, TValue, TFilter>(name, (Func<TKey, TValue, TFilter>)filterLambda.Compile());
+        }
+
 		/// <summary>
 		/// Retrieves all keys that match the given filter value, or an empty array if there are no matches.
 		/// </summary>
-		public async Task<IEnumerable<TKey>> FilterAsync(ITransaction tx, TFilter filter, int count, TimeSpan timeout, CancellationToken token)
+		public async Task<IEnumerable<TKey>> FilterAsync(ITransaction tx, TFilter filter, TimeSpan timeout, CancellationToken token)
 		{
 			var result = await _index.TryGetValueAsync(tx, filter, timeout, token).ConfigureAwait(false);
-			return result.HasValue ? result.Value.Take(count) : Enumerable.Empty<TKey>();
+			TKey[] results =  result.HasValue ? result.Value : new TKey[] { };
+            Array.Sort(results);
+            return results.AsEnumerable();
 		}
 
 		/// <summary>
-		/// Retrieves all keys that fall in the given filter range (inclusively), or an empty array if there are no matches.
+		/// Retrieves all keys that fall in the given filter range, or an empty array if there are no matches.
 		/// </summary>
-		public async Task<IEnumerable<TKey>> RangeFilterAsync(ITransaction tx, TFilter start, TFilter end, int count, CancellationToken token)
+		public async Task<IEnumerable<TKey>> RangeFilterAsync(ITransaction tx, TFilter start, RangeFilterType startType, TFilter end, RangeFilterType endType, CancellationToken token)
 		{
-			// Since filters uses exact matches, each key should appear exactly once in the index.
-			var keys = new List<TKey>();
+            if (startType == RangeFilterType.INCLUSIVE && endType == RangeFilterType.INCLUSIVE)
+            {
+                return await RangeFilterHelper(tx, f => start.CompareTo(f) <= 0 && f.CompareTo(end) <= 0, token);
```

Comment:
Since RangeFilterHelper uses index's CreatEnumerableAsync() doesn't that mean RangeFilterHelper has to be async and bc of that RangeFilter also async?

Reply:
It will still be async (Task return type indicates an async/awaitable method).  These two definitions below are exactly equivalent, but the second one is more performant (in cases where the compiler doesn't optimize):

public async Task<int> GetAsync() { return await GetHelperAsync(); }
public Task<int> GetAsync() { return GetHelperAsync(); }

Reply:
Oh I see, as in don't await something until you actually need the result